draft-bbs-signatures.md

```diff
@@ -680,38 +680,56 @@ Procedure:
 
 # Utility Operations
 
-## CreateGenerators
+## Generator point computation
 
-The CreateGenerators operation defines how to create a set of generators that form a part of the public parameters used by the BBS Signature scheme to accomplish operations such as Sign, Verify, ProofGen and ProofVerify.
+The `create_generators` operation defines how to create a set of generators that form a part of the public parameters used by the BBS Signature scheme to accomplish operations such as Sign, Verify, ProofGen and ProofVerify. It takes one input, the number of generator points to create, which is determined in part by the number of signed messages.
+
+As an optimization, implementations MAY cache the result of `create_generators` for a specific `generator_seed` (determined by the ciphersuite) and `count`. The values `n` and `v` MAY also be cached in order to efficiently extend a existing list of generator points.
 
 ```
-generators = CreateGenerators(dst, message_generator_seed, length);
+generators = create_generators(count)
 
 Inputs:
 
-- dst (REQUIRED), octet string. Domain Separation Tag.
-- message_generator_seed (REQUIRED), octet string.
-- length (REQUIRED), unsigned integer. Number of generators to create from the seed and dst.
+- count (REQUIRED), unsigned integer. Number of generator points to create.
+
+Parameters:
+
+- generator_seed, octet string. A seed value selected by the ciphersuite.
+- hash_to_curve_g1. A hash-to-curve suite selected by the ciphersuite.
+
+Definitions:
+
+- expand_message is the corresponding operation associated with the ciphersuite's hash_to_curve_g1 parameter.
+- seed_dst is the octet string representing "BBS-SIG-GENERATOR-SEED" in the ASCII character encoding.
+- hash_output_bytes is the integer given by ceil((log2(r) + k) / 8).
+- generator_dst is the octet string representing "BBS-SIG-GENERATOR-DST" in the ASCII character encoding.
 
 Outputs:
 
 - generators, an array of generators.
 
 Procedure:
 
-1. h = xof(seed)
+1. v = expand_message(generator_seed, seed_dst, hash_output_bytes)
+
+2. n = 1
+
+3. for i in range(1, count):
 
-2. for i in 0 to length:
+4.    generator_i = Identity_G1
 
-3.    generator_i = Identity_G1
+5.    while generator_i == Identity_G1 or generator_i == P1:
 
-4.    while(generator_i == Identity_G1 or generator_i == P1)
+6.        v = expand_message(v || I2OSP(n, 4), seed_dst, hash_output_bytes)
 
-5.        candidate = hash_to_curve_g1(h.read(xof_no_of_bytes), dst)
+7.        n = n + 1
 
-6.        if candidate not in generators: generator_i = candidate
+8.        candidate = hash_to_curve_g1(v, generator_dst)
 
-7. return generators
+9.        if candidate not in generators[1..i-1]: generator_i = candidate
+
+10. return generators
```

Comment:
```suggestion
9.        if candidate not in (generator_1, ..., generator_(i-1)):

10.              generator_i = candidate

11. return (generator_1, ..., generator_n)
```
Just a suggestion because we "technically" don’t define the generators list

Reply:
Thanks, maybe it should use `n` instead of `count` as the length parameter in line with hash-to-scalar?

Reply:
Oh right, sorry! Yea I’m fine with both honestly. I'm updating hash-to-scalar either way, so I can make that consistent with what we do here. I updated the suggestion to use `count` for now, just for ease, but I can go with either.